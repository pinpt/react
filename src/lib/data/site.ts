import type { Site } from '../types/site';
import { executeAPI } from '../fetch';

export const fetchSite = async (siteId: string): Promise<Site> => {
	const { site } = await executeAPI(`/site-api/v1/site/${siteId}`);
	return site;
};

type AnalyticsResult = Record<string, { claps: number; pageviews: number }>;

export const fetchAnalytics = async (siteId: string): Promise<AnalyticsResult> => {
	const { data } = await executeAPI(`/site-api/v1/site/${siteId}/analytics`);
	return data;
};
