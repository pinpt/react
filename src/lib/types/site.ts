export interface ISite {
	id: string;
	logoUrl?: string;
	mailingAddress?: string;
	url: string; // the canonical url for the site
	slug: string;
	name: string;
	private: boolean; // true if private and should not be indexable by search engines
	theme?: {
		title?: string;
		social?: Record<string, string>;
		description?: string;
		logoLink?: string;
		copyright?: string;
		eulaLink?: string;
		privacyLink?: string;
		tosLink?: string;
		homepage?: {
			tags?: string[];
		};
	};
	scriptPlugins?: Record<string, any>;
	basePath: string;
}
