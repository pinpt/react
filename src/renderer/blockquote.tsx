import { registerNode, recurseIntoChildren, NodeProps } from './register';

const BlockQuote = ({ node }: NodeProps) => {
	return <blockquote>{recurseIntoChildren(node)}</blockquote>;
};

registerNode('blockquote', (node) => <BlockQuote node={node} />);
