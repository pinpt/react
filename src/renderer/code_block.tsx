import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { twilight } from 'react-syntax-highlighter/dist/cjs/styles/prism';
import { NodeProps, registerNode } from './register';

const CodeBlock = ({ node }: NodeProps) => {
	return (
		<SyntaxHighlighter
			language={node.attrs.language}
			style={twilight}
			customStyle={{
				background: 'black',
				fontFamily: 'ui-monospace',
				borderWidth: '0px',
				borderRadius: '.25em',
			}}
		>
			{node.content?.[0].text ?? ''}
		</SyntaxHighlighter>
	);
};

registerNode('code_block', (node) => <CodeBlock node={node} />);
