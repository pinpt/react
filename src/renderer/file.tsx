import {
	faDownload, faFile, faFileAlt, faFileArchive, faFileAudio, faFileCode, faFileExcel, faFilePdf,
	faFilePowerpoint, faFileVideo, faFileWord
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { NodeProps, registerNode } from './register';

export const iconForType = (type = '') => {
	if (/pdf/.test(type)) {
		return faFilePdf;
	}
	if (/zip|tar/.test(type)) {
		return faFileArchive;
	}
	if (/video/.test(type)) {
		return faFileVideo;
	}
	if (/audio/.test(type)) {
		return faFileAudio;
	}
	if (/excel|spreadsheet/.test(type)) {
		return faFileExcel;
	}
	if (/powerpoint/.test(type)) {
		return faFilePowerpoint;
	}
	if (/word/.test(type)) {
		return faFileWord;
	}
	if (/css|csv|graphql|html|java|json|php|python|typescript|xml/.test(type)) {
		return faFileCode;
	}
	if (/text/.test(type)) {
		return faFileAlt;
	}
	return faFile;
};

export const formattedSize = (size = 0) => {
	const kB = size / 1000;
	if (kB < 10) {
		return `${Math.round(kB * 10) / 10} KB`;
	}
	if (kB < 1000) {
		return `${Math.round(kB)} KB`;
	}
	const MB = kB / 1000;
	return `${Math.round(MB * 10) / 10} MB`;
};

const File = ({ node }: NodeProps) => {
	const { description, filename, publicUrl, size, type } = node.attrs;

	const onDownload = () => {
		if (publicUrl) {
			const u = new URL(publicUrl);
			u.searchParams.set('filename', filename);
			u.searchParams.set('disposition', 'attachment');
			const a = document.createElement('a');
			a.style.display = 'none';
			a.href = u.toString();
			// the filename you want
			a.download = filename;
			document.body.appendChild(a);
			a.click();
			a.remove();
		}
	};

	return (
		<div className="file">
			<div className="container">
				<FontAwesomeIcon icon={iconForType(type)} />
				<div className="details">
					{publicUrl ? (
						<a href={publicUrl} target="_blank" rel="noopener noreferrer" className="filename">
							{filename}
						</a>
					) : (
						<div className="filename">{filename}</div>
					)}
					{description && <div className="description">{description}</div>}
					<div className="size">{formattedSize(size)}</div>
				</div>
				<div className="options">
					{publicUrl && (
						<div className="option group" onClick={onDownload}>
							<FontAwesomeIcon icon={faDownload} />
							<span>Download</span>
						</div>
					)}
				</div>
			</div>
		</div>
	);
};

registerNode('file', (node) => <File node={node} />);
