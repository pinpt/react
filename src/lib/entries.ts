import { Entry } from './types';

export const splitEntries = (entries: Entry[] = [], latestCount = 1) => {
	const latest = entries.slice(0, latestCount);
	const recent = entries.slice(latestCount, entries.length);

	return { latest, recent };
}