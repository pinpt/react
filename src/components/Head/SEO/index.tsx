import type { ISite, IContent } from '../../../lib/types';
import React from 'react';
import { extractImageMetadataFromFileID } from '../../../lib/file_metadata';
import { getTwitterProfileFromURL } from '../../../lib/string';

export interface ISEOProps {
	site: ISite;
	content?: IContent;
}

const SiteSEO = ({ site }: { site: ISite }) => {
	const title = site.theme?.description ? `${site.theme.description} - ${site.name}` : site.name;
	const description = site.theme?.description ?? site.name;
	return (
		<>
			<title>{title}</title>
			<meta name="type" property="og:type" content="website" />
			<meta name="title" property="og:title" content={title} />
			<meta name="twitter:title" content={title} />
			<meta name="description" property="og:description" content={description} />
			<meta name="twitter:description" content={description} />
		</>
	);
};

const formatAuthors = (content: IContent): string => {
	const authors =
		content.authors?.filter((author) => !!author).map((author) => `${author.firstName} ${author.lastName}`) ?? [];
	return authors.join(', ');
};

const getCoverImage = (content: IContent): string => {
	if (content.coverMedia?.placeholderImage) {
		const { size } = extractImageMetadataFromFileID(content.coverMedia.placeholderImage);
		if (size?.width !== 1200 && size?.height !== 628) {
			const u = new URL(content.coverMedia.placeholderImage);
			u.searchParams.set('rw', '1200');
			u.searchParams.set('rh', '628');
			return u.toString();
		}
		return content.coverMedia.placeholderImage;
	}
	return '';
};

const ContentSEO = ({ site, content }: { site: ISite; content: IContent }) => {
	const title = content.title;
	const description = content.headline;
	const author = formatAuthors(content);
	const date = new Date(content.publishedAt).toISOString();
	const coverImage = getCoverImage(content);
	return (
		<>
			<title>
				{title} - {site.name}
			</title>
			{coverImage && (
				<>
					<meta name="image" property="og:image" content={coverImage} />
					<meta name="twitter:image" content={coverImage} />
					<meta name="twitter:image:alt" content={title} />
				</>
			)}
			<meta name="type" property="og:type" content="article" />
			<meta property="author" content={author} />
			<meta name="author" property="og:author" content={author} />
			<meta property="article:published_time" content={date} />
			<meta name="title" property="og:title" content={title} />
			<meta name="twitter:title" content={title} />
			<meta name="description" property="og:description" content={description} />
			<meta name="twitter:description" content={description} />
			<link
				rel="alternate"
				type="application/rss+xml"
				title={`RSS Feed for ${site.theme?.title ?? site.name}`}
				href={`${site.url}/rss`}
			/>
		</>
	);
};

const SEO = (props: ISEOProps) => {
	const { site, content } = props;
	const contentSEO = content ? <ContentSEO site={site} content={content} /> : null;
	const siteSEO = !content ? <SiteSEO site={site} /> : null;
	const url = content ? content.url : site.url;
	const twitterHandle = getTwitterProfileFromURL(site.theme?.social?.twitter);
	return (
		<>
			{contentSEO}
			{siteSEO}
			<link rel="icon" href={site.logoUrl} />
			<link rel="shortcut icon" href={site.logoUrl} />
			<link rel="alternate" type="application/rss+xml" title={`${site.name} RSS Feed`} href="/rss" />
			<meta property="og:site_name" content={site.name} />
			<meta name="robots" content="index, follow" />
			<meta name="twitter:card" content="summary_large_image" />
			<meta property="og:url" content={url} />
			<meta property="twitter:url" content={url} />
			<link rel="canonical" href={url} />
			{twitterHandle && (
				<>
					<meta name="twitter:creator" content={twitterHandle} />
					<meta name="twitter:site" content={twitterHandle} />
				</>
			)}
		</>
	);
};

export default SEO;
