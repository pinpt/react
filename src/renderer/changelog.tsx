import { ReactYouTubeLite as Youtube } from 'react-youtube-lite';
import { getFileUrl } from '../lib/file_metadata';
import { slugifyContent } from '../lib/string';
import { ChangelogCoverMedia, ChangelogMediaType } from '../lib/types';
import { Content as Doc } from './';

const ImageMedia = ({ src, title = '' }: { src: string; title?: string }) => {
	return (
		<div className="image">
			<img src={src} alt={title} />
		</div>
	);
};

const VideoMedia = ({ src }: { src: string }) => {
	return (
		<div className="video">
			<video controls src={src} />
		</div>
	);
};

const YoutubeMediaSSR = ({ id, poster = 'hqdefault' }: { id: string; poster?: string }) => {
	// the component isn't SSR friendly so this is basically used to sort of mimic it for generated code
	const posterUrl = `https://i.ytimg.com/vi/${id}/${poster}.jpg`;
	return (
		<>
			{poster && <link rel="preload" href={posterUrl} as="image" />}
			<div
				className="youtube ryt-lite embed-responsive aspect-ratio-16/9"
				style={{ backgroundImage: `url(${posterUrl})` }}
			>
				<div className="lty-playbtn" data-url={`https://www.youtube-nocookie.com/embed/${id}?autoplay=1`} />
			</div>
		</>
	);
};

const YoutubeMedia = ({
	id,
	metadata,
	staticMode,
}: {
	id: string;
	metadata?: Record<string, any>;
	staticMode?: true;
}) => {
	return (
		<div className="youtube">
			{!staticMode ? (
				<Youtube url={`https://www.youtube.com/watch?v=${id}`} poster={metadata?.poster} />
			) : (
				<YoutubeMediaSSR id={id} poster={metadata?.poster} />
			)}
		</div>
	);
};

const Media = ({ media, title, staticMode }: { media?: ChangelogCoverMedia; title?: string; staticMode?: true }) => {
	if (!media) {
		return <></>;
	}
	let content: React.ReactElement;
	switch (media.type) {
		case ChangelogMediaType.None: {
			return <></>;
		}
		case ChangelogMediaType.Image: {
			content = <ImageMedia src={getFileUrl(media.value)} title={title} />;
			break;
		}
		case ChangelogMediaType.Video: {
			content = <VideoMedia src={getFileUrl(media.value)} />;
			break;
		}
		case ChangelogMediaType.Youtube: {
			content = <YoutubeMedia id={media.value} metadata={media.metadata} staticMode={staticMode} />;
			break;
		}
	}
	return <div className="media-container">{content}</div>;
};

interface ChangelogProps {
	id: string;
	title?: string;
	content?: any;
	coverMedia?: ChangelogCoverMedia;
	limit?: number;
	divider?: boolean;
	staticMode?: true;
}

const ChangelogLog = (props: ChangelogProps) => {
	return (
		<article className="changelog notebook-editor read-only">
			{props.coverMedia && (
				<section className="covermedia">
					<Media media={props.coverMedia} title={props.title} staticMode={props.staticMode} />
				</section>
			)}
			<section className="ProseMirror">
				<Doc node={props.content} limit={props.limit} />
				{props.limit && props.content.content?.length > props.limit && (
					<div className="continue">
						<a href={slugifyContent(props.id, props.title)}>Continue Reading â†’</a>
					</div>
				)}
			</section>
			{props.divider && <hr className="divider" />}
		</article>
	);
};

export default ChangelogLog;
