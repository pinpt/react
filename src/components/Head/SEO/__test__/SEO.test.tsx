import renderer from 'react-test-renderer';
import SEO from '../';
import entries from '../../../Prebuilt/__data__/testEntries.json';
import site from '../../../Prebuilt/__data__/testSite.json';

import type { IContent } from '../../../../lib/types';

test('Test default render', () => {
	const component = renderer.create(<SEO site={site} content={entries[0] as IContent} />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
