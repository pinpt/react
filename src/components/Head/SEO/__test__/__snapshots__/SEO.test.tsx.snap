// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test default render 1`] = `
Array [
  <title>
    Test
     - 
    Jeff's Testing Site -- NONE OF THIS INFO IS REAL
  </title>,
  "",
  <meta
    content="article"
    name="type"
    property="og:type"
  />,
  <meta
    content="Jeff Haynie"
    property="author"
  />,
  <meta
    content="Jeff Haynie"
    name="author"
    property="og:author"
  />,
  <meta
    content="2021-08-03T21:47:06.965Z"
    property="article:published_time"
  />,
  <meta
    content="Test"
    name="title"
    property="og:title"
  />,
  <meta
    content="Test"
    name="twitter:title"
  />,
  <meta
    content="Hi there. What’s happening? Just testing some stuff here."
    name="description"
    property="og:description"
  />,
  <meta
    content="Hi there. What’s happening? Just testing some stuff here."
    name="twitter:description"
  />,
  <link
    href="https://file.edge.pinpoint.com/b24dabeec1f19705f31ed39278729729;UgK7%23L%2C-n%25ay%7D7bbR-aft2J-Naa%23%23rW-NwWW;128x129.png"
    rel="icon"
  />,
  <link
    href="https://file.edge.pinpoint.com/b24dabeec1f19705f31ed39278729729;UgK7%23L%2C-n%25ay%7D7bbR-aft2J-Naa%23%23rW-NwWW;128x129.png"
    rel="shortcut icon"
  />,
  <link
    href="/rss"
    rel="alternate"
    title="Jeff's Testing Site -- NONE OF THIS INFO IS REAL RSS Feed"
    type="application/rss+xml"
  />,
  <meta
    content="Jeff's Testing Site -- NONE OF THIS INFO IS REAL"
    property="og:site_name"
  />,
  <meta
    content="index, follow"
    name="robots"
  />,
  <meta
    content="summary_large_image"
    name="twitter:card"
  />,
  <meta
    content="https://jeff.pinpoint.com/entry/14Y3mpvdnQKf45Wtue0t/Test"
    property="og:url"
  />,
  <meta
    content="https://jeff.pinpoint.com/entry/14Y3mpvdnQKf45Wtue0t/Test"
    property="twitter:url"
  />,
  <link
    href="https://jeff.pinpoint.com/entry/14Y3mpvdnQKf45Wtue0t/Test"
    rel="canonical"
  />,
  <meta
    content="@jhaynie"
    name="twitter:creator"
  />,
  <meta
    content="@jhaynie"
    name="twitter:site"
  />,
]
`;
